AABB_tree/include/CGAL/AABB_segment_primitive_2.h

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2012 INRIA Sophia-Antipolis (France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0-or-later OR LicenseRef-Commercial
+//
+//
+// Author(s)     : Sebastien Loriot
+//
+
+
+#ifndef CGAL_AABB_SEGMENT_PRIMITIVE_2_H_
+#define CGAL_AABB_SEGMENT_PRIMITIVE_2_H_
+
+#include <CGAL/license/AABB_tree.h>
+
+#include <CGAL/disable_warnings.h>
+
+#include <CGAL/AABB_primitive.h>
+#include <iterator>
+
+namespace CGAL {
+
+namespace internal {
+  template <class GeomTraits, class Iterator>
+  struct Source_of_segment_2_iterator_property_map{
+    //classical typedefs
+    typedef Iterator key_type;
+    typedef typename GeomTraits::Point_2 value_type;
+    // typedef decltype(
+    //   std::declval<typename GeomTraits::Construct_source_2>()(
+    //     std::declval<typename GeomTraits::Segment_2>())) reference;
+    typedef decltype(
+      typename GeomTraits::Construct_source_2()(
+        *std::declval<key_type&>())) reference;
+    typedef boost::readable_property_map_tag category;
+    typedef Source_of_segment_2_iterator_property_map<GeomTraits, Iterator> Self;
+
+    inline friend reference
+    get(Self, key_type it)
+    {
+      return typename GeomTraits::Construct_source_2()( *it );
+    }
+  };
+}//namespace internal
+
+
+/*!
+ * \ingroup PkgAABBTreeRef
+ * Primitive type that uses as identifier an iterator with a 2D segment as `value_type`.
+ * The iterator from which the primitive is built should not be invalided
+ * while the AABB tree holding the primitive is in use.
+ *
+ * \cgalModels{AABBPrimitive}
+ *
+ * \tparam GeomTraits is a traits class providing the nested type `Point_2` and `Segment_2`.
+ *         It also provides the functor `Construct_source_2` that has an operator taking a `Segment_2`
+ *         and returning its source as a type convertible to `Point_2`.
+ * \tparam Iterator is a model of `ForwardIterator` with its value type convertible to `GeomTraits::Segment_2`
+ * \tparam CacheDatum is either `CGAL::Tag_true` or `CGAL::Tag_false`. In the former case,
+ *           the datum is stored in the primitive, while in the latter it is
+ *           constructed on the fly to reduce the memory footprint.
+ *           The default is `CGAL::Tag_false` (datum is not stored).
+ *
+ * \sa `AABBPrimitive`
+ * \sa `AABB_primitive<Id,ObjectPropertyMap,PointPropertyMapPolyhedron,ExternalPropertyMaps,CacheDatum>`
+ * \sa `AABB_segment_primitive_3<Iterator,CacheDatum>`
+ * \sa `AABB_triangle_primitive_2<Iterator,CacheDatum>`
+ * \sa `AABB_triangle_primitive_3<Iterator,CacheDatum>`
+ * \sa `AABB_halfedge_graph_segment_primitive<HalfedgeGraph,OneHalfedgeGraphPerTree,CacheDatum>`
+ * \sa `AABB_face_graph_triangle_primitive<FaceGraph,OneFaceGraphPerTree,CacheDatum>`
```

Comment:
Can probably purge a bit the \sa, what's the link between Segment_2 primitive and 3D FaceGraph primitive?